{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "everyday-dayton",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Args\n",
    "import argparse\n",
    "import time\n",
    "\n",
    "# Get args\n",
    "def get_args():\n",
    "    parser = argparse.ArgumentParser(description = \"Predict diverse RNA subcellular localization with RNA sequence (cDNA format: ACGT) and identify determinant k-mers with Shapely value\", usage = \"RNAlight -q query.fasta [-p Prefix] [-o Output_Dir] [--shap Shapely_value] [--topn Top_n_k-mer] [-m mRNA] [--RNA RNA] [-h] [-v]\", epilog=\"Happy Light!\")\n",
    "    parser.add_argument(\"-q\",\"--query\", dest = \"query\", type = str, metavar=(\"Query\"),required = True, help = \"query file to be predicted (fasta format)\")\n",
    "    parser.add_argument(\"-p\",\"--prefix\",dest = \"prefix\", type = str, default = \"query\", metavar=(\"Prefix\"), help = \"prefix of outputfiles (default: query)\")\n",
    "    parser.add_argument(\"-o\",\"--outputdir\",dest = \"outputdir\",type = str, default = \"./RNAlight_output\", metavar=(\"Output_Dir\"), help = \"output directions of outputfiles (default:RNAlight_output)\")\n",
    "    parser.add_argument(\"--shap\",dest = \"shap\",action = \"store_true\", help = \"Provide determinant k-mers using Shapely value\")\n",
    "    parser.add_argument(\"--topn\",dest = \"topn\",type = int,default = 10, help = \"Top n determinant k-mers computed by Shaply value (default: 10)\")\n",
    "    parser.add_argument(\"-m\",\"--mRNA\",dest = \"mRNA\", action = \"store_true\", help = \"query sequence is mRNA\")\n",
    "    parser.add_argument(\"--RNA\",dest = \"RNA\",action = \"store_true\", help = \"query sequence is RNA format(ACGU)\")\n",
    "    parser.add_argument(\"-v\",\"--version\",action=\"version\",version='RNAlight v1.0',help = \"version for RNAlight\")\n",
    "    args = parser.parse_args()\n",
    "    return args\n",
    "\n",
    "\n",
    "args = get_args()\n",
    "query_file = args.query\n",
    "mRNA = args.mRNA\n",
    "RNA = args.RNA\n",
    "s_shap = args.shap\n",
    "topn = args.topn\n",
    "prefix = args.prefix\n",
    "outputdir = args.outputdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "loaded-aruba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python import\n",
    "print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]:\",\"\\tLoading requirements......\\n\")\n",
    "import os\n",
    "import copy\n",
    "import itertools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.externals import joblib\n",
    "import shap\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# Count the frequency of k-mer in each RNA sequence\n",
    "# k-mer was normalized by total k-mer count of each RNA sequence\n",
    "def _count_kmer(Dataset,k): # k = 3,4,5\n",
    "    \n",
    "    # copy dataset\n",
    "    dataset = copy.deepcopy(Dataset)\n",
    "    # alphbet of nucleotide\n",
    "    nucleotide = ['A','C','G','T']\n",
    "    \n",
    "    # generate k-mers\n",
    "    #  k == 5:\n",
    "    five = list(itertools.product(nucleotide,repeat=5))\n",
    "    pentamer = []\n",
    "    for n in five:\n",
    "        pentamer.append(\"\".join(n))\n",
    "    \n",
    "    #  k == 4:\n",
    "    four = list(itertools.product(nucleotide,repeat=4))\n",
    "    tetramer = []\n",
    "    for n in four:\n",
    "        tetramer.append(\"\".join(n))\n",
    "\n",
    "    # k == 3:\n",
    "    three = list(itertools.product(nucleotide,repeat=3))\n",
    "    threemer = []\n",
    "    for n in three:\n",
    "        threemer.append(\"\".join(n))\n",
    "    \n",
    "    # input features can be combinations of diffrent k values\n",
    "    if k == 34:\n",
    "        table_kmer = dict.fromkeys(threemer,0)\n",
    "        table_kmer.update(dict.fromkeys(tetramer,0))\n",
    "    if k == 45:\n",
    "        table_kmer = dict.fromkeys(tetramer,0)\n",
    "        table_kmer.update(dict.fromkeys(pentamer,0))\n",
    "    if k == 345:\n",
    "        table_kmer = dict.fromkeys(threemer,0)\n",
    "        table_kmer.update(dict.fromkeys(tetramer,0))\n",
    "        table_kmer.update(dict.fromkeys(pentamer,0))\n",
    "\n",
    "    # count k-mer for each sequence\n",
    "    for mer in table_kmer.keys():\n",
    "        table_kmer[mer] = dataset[\"cdna\"].apply(lambda x : x.count(mer))\n",
    "    \n",
    "    # for k-mer raw count without normalization\n",
    "    rawcount_kmer_df = pd.DataFrame(table_kmer)\n",
    "    df1_rawcount = pd.concat([rawcount_kmer_df,dataset[\"seq_id\"]],axis = 1)\n",
    "\n",
    "\n",
    "    # for k-mer frequency with normalization\n",
    "    freq_kmer_df = rawcount_kmer_df.apply(lambda x: x/x.sum(),axis=1)\n",
    "    df1 = pd.concat([freq_kmer_df,dataset[\"seq_id\"]],axis = 1)\n",
    "\n",
    "    return df1,df1_rawcount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "catholic-exploration",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output dir\n",
    "if not (os.path.exists(outputdir)):\n",
    "    os.mkdir(outputdir)\n",
    "\n",
    "# tmp_dir\n",
    "tmp_dir_name = prefix + \".tmp\"\n",
    "tmp_dir = os.path.join(outputdir,tmp_dir_name)\n",
    "if not (os.path.exists(tmp_dir)):\n",
    "    os.mkdir(tmp_dir)\n",
    "\n",
    "# Convert fasta file to tab format\n",
    "tab_tmp = prefix + \".txt\"\n",
    "if RNA:\n",
    "    fa2tab = \"seqkit seq --dna2rna \" + query_file +\" | seqkit fx2tab > \" + os.path.join(tmp_dir,tab_tmp)\n",
    "else:\n",
    "    fa2tab = \"seqkit fx2tab \" + query_file + \" > \" + os.path.join(tmp_dir,tab_tmp)\n",
    "\n",
    "os.system(fa2tab)\n",
    "\n",
    "# +\n",
    "# Load RNAlight model\n",
    "print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tStart Loading RNAlight......\\n\")\n",
    "\n",
    "\n",
    "absolute_path = os.popen(\"dirname `which RNAlight`\").read().strip('\\n')\n",
    "\n",
    "if mRNA:\n",
    "    RNA_Light = joblib.load(os.path.join(absolute_path,\"src/RNAlight_model/RNAlight_mRNA_model.pkl\"))\n",
    "else:\n",
    "    RNA_Light = joblib.load(os.path.join(absolute_path,\"src/RNAlight_model/RNAlight_lncRNA_model.pkl\"))\n",
    "\n",
    "\n",
    "print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tSuccessfully Loaded RNAlight!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "violent-screen",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data Processing\n",
    "print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]:\",\"\\tStart Data Processing......\\n\")\n",
    "\n",
    "\n",
    "query = pd.read_csv(os.path.join(tmp_dir,tab_tmp),sep='\\t',index_col = False, names = [\"seq_id\",\"cdna\"])\n",
    "df_kmer_345,df_kmer_345_rawcount = _count_kmer(query,345)\n",
    "\n",
    "kmer_file = prefix + \"_df_kmer345_freq.tsv\"\n",
    "kmer_rowcount_file = prefix + \"_df_kmer345_rawcount.tsv\"\n",
    "df_kmer_345.to_csv(os.path.join(outputdir,kmer_file),sep='\\t')\n",
    "df_kmer_345_rawcount.to_csv(os.path.join(outputdir,kmer_rowcount_file),sep='\\t')\n",
    "\n",
    "\n",
    "print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tSuccessfully Finished Data Processing!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "empty-johnson",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-09-24 09:36:43 \tStart Predicting......\n",
      "\n",
      "2021-09-24 09:36:43 \tSuccessfully Finished Predicting!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Predict \n",
    "print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tStart Predicting......\\n\")\n",
    "\n",
    "\n",
    "del df_kmer_345['seq_id']\n",
    "x_kmer = df_kmer_345.values\n",
    "\n",
    "y_pred = RNA_Light.predict(x_kmer)\n",
    "y_prob = RNA_Light.predict_proba(x_kmer)[:,1]\n",
    "\n",
    "# Output\n",
    "query[\"RNALight_pred_label\"] =  y_pred\n",
    "query[\"RNALight_pred_prob\"] = y_prob\n",
    "query[\"Light_score\"] = 2*y_prob-1 \n",
    "\n",
    "outputfile = prefix + \"_RNAlight_predict_df.txt\"\n",
    "query.to_csv(os.path.join(outputdir,outputfile),sep = '\\t',index=False)\n",
    "\n",
    "\n",
    "print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tSuccessfully Finished Predicting!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "foster-therapist",
   "metadata": {},
   "outputs": [],
   "source": [
    "if s_shap:\n",
    "    print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tStart Identifying determinant k-mers......\\n\")\n",
    "    #shap.initjs() \n",
    "    # Identify determinant k-mers with Shapely value\n",
    "    RNA_Light_explainer = shap.TreeExplainer(RNA_Light)\n",
    "    shap_values_RNA_Light = RNA_Light_explainer.shap_values(df_kmer_345)[1]\n",
    "    \n",
    "    shap_df = pd.DataFrame(shap_values_RNA_Light,columns = df_kmer_345.columns)\n",
    "    shap_df.index = query[\"seq_id\"]\n",
    "    output_shap_df = prefix + \"_shap_df.txt\"\n",
    "    shap_df.to_csv(os.path.join(outputdir,output_shap_df),sep = '\\t')\n",
    "    \n",
    "    # Define colormap\n",
    "    my_colormap= plt.get_cmap(\"RdBu_r\")\n",
    "\n",
    "    # Plot the summary without showing it\n",
    "    plt.figure()\n",
    "    # Violin plot\n",
    "    shap.summary_plot(shap_values_RNA_Light,features =df_kmer_345,max_display = topn,show=False)\n",
    "\n",
    "    # Change the colormap of the artists\n",
    "    for fc in plt.gcf().get_children():\n",
    "        for fcc in fc.get_children():\n",
    "            if hasattr(fcc, \"set_cmap\"):\n",
    "                fcc.set_cmap(my_colormap)\n",
    "    output_shap_violin_plot = prefix + \"_shap_violin_plot_top\" + str(topn) +\"_k-mer.pdf\"\n",
    "    plt.savefig(os.path.join(outputdir,output_shap_violin_plot), format='pdf', dpi=1200)\n",
    "    \n",
    "    # Bar plot\n",
    "    plt.figure()\n",
    "    shap.summary_plot(shap_values_RNA_Light,df_kmer_345,plot_type = \"bar\",max_display = topn,color = \"#8d99ae\",show = False)\n",
    "    output_shap_bar_plot = prefix + \"_shap_bar_plot_top\" + str(topn) +\"_k-mer.pdf\"\n",
    "    plt.savefig(os.path.join(outputdir,output_shap_bar_plot), format='pdf', dpi=1200)\n",
    "    print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tSuccessfully Finished Identifying determinant k-mers!\\n\")\n",
    "    print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tSuccessfully Finished RNAlight!\")\n",
    "else:\n",
    "    print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\t\\nProcedure for Identifying determinant k-mers were not executed. Please add --shap if you need.\")\n",
    "    print(\"[\",time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()),\"]\",\"\\tSuccessfully Finished RNAlight!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "future-manufacturer",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch",
   "language": "python",
   "name": "pytorch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
