{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "chicken-consensus",
   "metadata": {},
   "source": [
    "# Bash"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "established-latex",
   "metadata": {},
   "source": [
    "***\n",
    "###  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "crude-disclaimer",
   "metadata": {},
   "source": [
    "- quant.txt was generated by CLEAR(CIRCexplorer3) using PA1 cell riboâ€“ RNA-seq data (GEO:GSE73325) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "special-tourism",
   "metadata": {},
   "outputs": [],
   "source": [
    "# circRNA derived from back-spliced exons\n",
    "\n",
    "awk '$14==\"circRNA\"' 00_PA1_steady_ribominus_rep1_quant.txt > 01_PA1_steady_ribominus_rep1_circRNA.txt\n",
    "awk '$14==\"circRNA\"' 00_PA1_steady_ribominus_rep2_quant.txt > 01_PA1_steady_ribominus_rep2_circRNA.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "affecting-repository",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ciRNA derived from intron lariat\n",
    "\n",
    "awk '$14==\"ciRNA\"' 00_PA1_steady_ribominus_rep1_quant.txt > 01_PA1_steady_ribominus_rep1_ciRNA.txt\n",
    "awk '$14==\"ciRNA\"' 00_PA1_steady_ribominus_rep2_quant.txt > 01_PA1_steady_ribominus_rep2_ciRNA.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comfortable-victorian",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
